sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample09_Extensibility.md

```diff
@@ -0,0 +1,395 @@
+
+    internal PluginProcessor(string queueName, ServiceBusClient client, IEnumerable<Func<ServiceBusReceivedMessage, Task>> plugins, ServiceBusProcessorOptions options) :
+        base(client, queueName, options)
+    {
+        _plugins = plugins;
+    }
+
+    internal PluginProcessor(string topicName, string subscriptionName, ServiceBusClient client, IEnumerable<Func<ServiceBusReceivedMessage, Task>> plugins, ServiceBusProcessorOptions options) :
+        base(client, topicName, subscriptionName, options)
+    {
+        _plugins = plugins;
+    }
+
+    protected internal override async Task OnProcessMessageAsync(ProcessMessageEventArgs args)
+    {
+        foreach (var plugin in _plugins)
+        {
+            await plugin.Invoke(args.Message);
+        }
+
+        await base.OnProcessMessageAsync(args);
+    }
+}
+```
+
+And a `PluginSessionProcessor`:
+```C# Snippet:End2EndPluginSessionProcessor
+string connectionString = "<connection_string>";
+string queueName = "<queue_name>";
+// since ServiceBusClient implements IAsyncDisposable we create it with "await using"
+await using var client = new ServiceBusClient(connectionString);
+await using ServiceBusSender sender = client.CreatePluginSender(queueName, new List<Func<ServiceBusMessage, Task>>()
+{
+    message =>
+    {
+        message.Subject = "Updated subject";
+        message.SessionId = "sessionId";
+        Console.WriteLine("First send plugin executed!");
+    return Task.CompletedTask;
+    },
+    message =>
+    {
+        Console.WriteLine(message.Subject); // prints "Updated subject"
+        Console.WriteLine(message.SessionId); // prints "sessionId"
+        Console.WriteLine("Second send plugin executed!");
+        return Task.CompletedTask;
+    },
+});
+
+await sender.SendMessageAsync(new ServiceBusMessage(Encoding.UTF8.GetBytes("First")));
+
+await using ServiceBusSessionProcessor processor = client.CreatePluginSessionProcessor(queueName, new List<Func<ServiceBusReceivedMessage, Task>>()
+{
+    message =>
+    {
+        var rawMessage = message.GetRawAmqpMessage();
+        rawMessage.Properties.Subject = "Received subject";
+        Console.WriteLine("First receive plugin executed!");
+        return Task.CompletedTask;
+    },
+    message =>
+    {
+        Console.WriteLine(message.Subject); // prints "Received subject"
+        var rawMessage = message.GetRawAmqpMessage();
+        rawMessage.Properties.Subject = "Last subject";
+        Console.WriteLine("Second receive plugin executed!");
+        return Task.CompletedTask;
+    },
+});
+
+processor.ProcessMessageAsync += args =>
+{
+    Console.WriteLine(args.Message.Subject);
+    return Task.CompletedTask;
+};
+
+processor.ProcessErrorAsync += args =>
+{
+    Console.WriteLine(args.Exception);
+    return Task.CompletedTask;
+};
+
+await processor.StartProcessingAsync();
+Console.ReadKey();
+```
+
+### Defining extension methods
```

Comment:
It takes a good while to get to this part, which I believe is one of the key ideas.   Have we considered changing the flow a bit to emphasize this?  Maybe doing this section first with just the `PluginSender1 to demonstrate the pattern?   Then we could describe the other sender/receiver/processor types and build on that knowledge?

Reply:
I think defining the types is actually the key part. There are a few ways to get the types hooked into the client. Extension methods are just one way.